contracts/voting-verifier/src/query.rs

```diff
@@ -1,28 +1,263 @@
+use axelar_wasm_std::voting::PollStatus;
 use connection_router::state::{CrossChainId, Message};
+use cosmwasm_schema::cw_serde;
 use cosmwasm_std::{Deps, StdResult};
 
 use crate::error::ContractError;
-use crate::state::VERIFIED_MESSAGES;
+use crate::state::{self, Poll, POLLS, POLL_MESSAGES};
 
-pub fn verification_statuses(
-    deps: Deps,
-    messages: Vec<Message>,
-) -> StdResult<Vec<(CrossChainId, bool)>> {
+#[cw_serde]
+pub enum VerificationStatus {
+    Verified,
+    FailedToVerify,
+    InProgress,  // still in an open poll
+    NotVerified, // not in a poll
+}
+
+pub fn is_verified(deps: Deps, messages: &[Message]) -> StdResult<Vec<(CrossChainId, bool)>> {
     messages
-        .into_iter()
+        .iter()
         .map(|message| {
-            is_message_verified(deps, &message).map(|verified| (message.cc_id, verified))
+            verification_status(deps, message).map(|status| match status {
+                VerificationStatus::Verified => (message.cc_id.to_owned(), true),
+                _ => (message.cc_id.to_owned(), false),
+            })
         })
         .collect::<Result<Vec<(_, _)>, _>>()
         .map_err(Into::into)
 }
 
-pub fn is_message_verified(deps: Deps, message: &Message) -> Result<bool, ContractError> {
-    match VERIFIED_MESSAGES.may_load(deps.storage, &message.cc_id)? {
-        Some(stored) if stored != *message => {
-            Err(ContractError::MessageMismatch(message.cc_id.to_string()))
+pub fn verification_status(
+    deps: Deps,
+    message: &Message,
+) -> Result<VerificationStatus, ContractError> {
+    match POLL_MESSAGES.may_load(deps.storage, &message.hash_id())? {
+        Some(stored) => {
+            let poll = POLLS
+                .load(deps.storage, stored.poll_id)
+                .expect("invalid invariant: message poll not found");
+
+            let verified = match &poll {
+                Poll::Messages(poll) | Poll::ConfirmWorkerSet(poll) => poll
+                    .has_consensus(stored.index_in_poll)
+                    .expect("invalid invariant: message not found in poll"),
+            };
+
+            if verified {
+                if stored.msg != *message {
+                    return Err(ContractError::MessageMismatch(message.cc_id.to_string()));
+                }
+                Ok(VerificationStatus::Verified)
+            } else if is_finished(&poll) {
+                Ok(VerificationStatus::FailedToVerify)
+            } else {
+                Ok(VerificationStatus::InProgress)
+            }
```

Comment:
```suggestion
            if verified {
                assert_eq!(stored.msg, *message, "...");

                return Ok(VerificationStatus::Verified);
            }

            if is_finished(&poll) {
                return Ok(VerificationStatus::FailedToVerify);
            }

            Ok(VerificationStatus::InProgress)
```
Avoid nested-if's and else and else if at all cost.

Reply:
yes to avoid nesting, but I prefer `else if` when the cases are on the same hierarchical level, i.e. `if... return` is fine when it's guard clauses, but when it's effectively a switch between multiple similar cases I want the code to reflect that

Reply:
agree with Chris, `else if` seem more readable in this case

Reply:
I never find else and else if readable and clean, but you guys insist sure,